core/src/docs/rfcs/3243_list_prefix.md

```diff
@@ -0,0 +1,128 @@
+| list not exist dir      | `def/`          | EMPTY                               | no children found matches the dir       |
+| list not exist file     | `def`           | EMPTY                               | no children found matches the prefix    |
+
+While listing a path with `delimiter` set to `""`:
+
+| Case                    | Path            | Result                                                                                        | Description                             |
+|-------------------------|-----------------|-----------------------------------------------------------------------------------------------|-----------------------------------------|
+| list dir                | `abc/`          | `abc/def_file` <br/> `abc/def_dir/` <br/> `abc/def_dir/xyz_file` <br/> `abc/def_dir/xyz_dir/` | children that matches the dir           |
+| list prefix             | `abc/def`       | `abc/def_file` <br/> `abc/def_dir/` <br/> `abc/def_dir/xyz_file` <br/> `abc/def_dir/xyz_dir/` | children that matches the prefix        |
+| list file               | `abc/def_file`  | `abc/def_file`                                                                                | the only children that matches the path |
+| list dir without `/`    | `abc/def_dir`   | `abc/def_dir/` <br/> `abc/def_dir/xyz_file` <br/> `abc/def_dir/xyz_dir/`                      | children that matches the path          |
+| list file ends with `/` | `abc/def_file/` | EMPTY                                                                                         | no children matches the dir             |
+| list not exist dir      | `def/`          | EMPTY                                                                                         | no children found matches the dir       |
+| list not exist file     | `def`           | EMPTY                                                                                         | no children found matches the prefix    |
+
+While stat a path:
+
+| Case                   | Path            | Result                                     |
+|------------------------|-----------------|--------------------------------------------|
+| stat existing dir      | `abc/`          | Metadata with dir mode                     | 
+| stat existing file     | `abc/def_file`  | Metadata with file mode                    | 
+| stat dir without `/`   | `abc/def_dir`   | Error `NotFound` or metadata with dir mode | 
+| stat file with `/`     | `abc/def_file/` | Error `NotFound`                           |
+| stat not existing path | `xyz`           | Error `NotFound`                           |
+
+While create dir on a path:
+
+| Case                        | Path   | Result                     |
+|-----------------------------|--------|----------------------------|
+| create dir on existing dir  | `abc/` | Ok                         |
+| create dir on existing file | `abc`  | Error with `NotADirectory` |
```

Comment:
Is this Result may be `FileExist`?

Reply:
OpenDAL only has `AlreadyExists`. Returning `AlreadyExists` may confuse our users since we do allow create dir on existing dir so I choose to return `NotADirectory` instead.

Reply:
`NotADirectory` is more confusing for me. It means that we try to list dir, and our input argument is not a directory.

Reply:
`NotADirectory` seems to make sense since we are calling `create_dir("abc")`. The error code clearly indicates that `abc is not a directory`. However, returning `already exists` is unclear and could give users the impression that the directory already exists.

Reply:
POSIX file system returns "file exists" for both `create_dir("abc/")` and `create_dir("abc")`. Therefore, the semantics could be different.

Reply:
May I ask why we allow creating dir on an existing dir?

Reply:
That make sense,  thanks!

Reply:
But if user calls create_dir("abc"), and gets an error of Not a Directory, isn't it also a little bit absurd? 

The user is creating a directory but getting an error of not a directory.

I think maybe we follow the posix api would be better? Since in regular file system we are getting an AlreadyExists Error on creating dir o an existing file.

Reply:
> But if user calls create_dir("abc"), and gets an error of Not a Directory, isn't it also a little bit absurd?
> 
> The user is creating a directory but getting an error of not a directory.

It does not a directory: the path is a file.

Reply:
> I guess the users are more likely to be familiar with the behavior of POSIX?

User should read opendal's docs. It's not a goal for opendal to behave like POSIX fs.

Reply:
Yea from the efficiency perspective that makes more sense.

Just a thought, If AlreadyExists is named by FileExists, does it works better?

Reply:
> Just a thought, If AlreadyExists is named by FileExists, does it works better?

No deep thought about that. The naming is inspired by HTTP Spec.